RFC-0023-Unified-Memory-for-Pytorch.md

```diff
@@ -0,0 +1,127 @@
+# Unified Memory for Pytorch
+
+This RFC proposes to add Unified Virtual Memory (UVM) (or “Managed Memory”) function. CUDA and ROCm include managed memory allocation APIs so this RFC applies to both. The terminology included references CUDA, but automatic hip conversion is expected to work via hipify.  Important usage and behavior differences in the managed memory APIs include control of memory allocations, device placement and DMA/copy scheduling. These APIs are also not limited to GPU memory sizes and offer a set of hints for control of the managed function.  
+
+The proposed changes to the front end and back end have been minimized as much as possible to have a very targeted effect when UVM is enabled and have no effect at all when UVM is disabled. 
+
+## Frontend Changes
+
+Minimal changes are proposed on the frontend as not to burden the user with changes outside of an enablement flag. Default settings for all new frontend additions will be off/disabled. UVM will not successfully enable when CUDA/ROCm is not present. 
+
+### Python
+
+We propose adding the following APIs to the python interface to enable and control the behavior of the UVM function: 
+
+
+`torch.cuda.set_enabled_uvm()` and `torch.cuda.get_enabled_uvm()`
+
+These APIs will be used to set and query unified memory usage in Pytorch. These will be defined in torch/cuda/memory.py. The `set_enabled_uvm()` API will set the enablement flag (enabled_uvm) which will be held in the default global context (`at::globalContext().userEnabledUVM()`). This location should allow us to avoid adding new parameters to functions that need to be modified. 
+
+The default value for enabled_uvm will be False. It is expected that when setting enabled_uvm to True, that it happens early in the Pytorch program before any allocations have taken place and then left on throughout. In order to enforce this, there is no proposed set_disabled_uvm() call. 
+
+`torch.cuda.set_uvm_move_on_copy()`
+
+This API will also set a flag (`enabled_uvm_move_on_copy`) in the globalContext. This flag will determine if tensor`.to()` should move the tensor to the new device instead of creating a copy. Aside from this flag, two other conditions must be met for a move to occur. 
+
+1. UVM is enabled 
+2. The `.to()` only changes the device and does not change any other tensor attribute. 
+
+`Torch.cuda.uvm_to(Tensor t)`
```

Comment:
We agree with your statement, and would like to propose using the 'copy' parameter in the 'to' function for a user to get the original behavior of the operation.
Such that tensor.to(device='cpu', copy=true) or the dtype or layout changes, we would force the tensor to allocate more storage and copy to the target device.

Otherwise, if the tensor.to() only changed the device type and copy=false(default), the operation would call a move_ function that uses the same memory, but changes DeviceType to the new device. 

A use case for this would be if you wanted a tensor to be accessible on different devices, but allowed to be changed asychronously.

Reply:
To illustrate, the following would always copy (assume the tensor is on cpu)
```
tensor.to(device='cpu', copy=true)
tensor.to(device='cuda:0', copy=false, dtype=torch.float32)
```
And this would call the move (zero copy)
```
tensor.to(device='cuda:0', copy=false)
```